test/connect.js

```diff
@@ -1,522 +1,450 @@
-test('reject client requested for unsupported protocol version', function (t) {
-  t.plan(3)
+test('reject client requested for unsupported protocol version', async (t) => {
+  t.plan(4)
 
-  Aedes.createBroker().then((broker) => {
-    t.teardown(broker.close.bind(broker))
+  const broker = await Aedes.createBroker()
+  t.after(() => broker.close())
+  const s = setup(broker)
 
-    const s = setup(broker)
+  // inStream.write throws an error when trying to write illegal packets
+  // so we use the encoded version:
+  const sendPacket = () => {
+    // cmd: 'connect', protocolId: 'MQTT', protocolVersion: 2, clean: true, clientId: 'my-client', keepalive: 0
+    const rawPacket = '10 15 00 04 4D 51 54 54 02 02 00 00 00 09 6D 79 2D 63 6C 69 65 6E 74'
+    rawWrite(s, rawPacket)
+  }
 
-    s.inStream.write({
-      cmd: 'connect',
-      protocolId: 'MQTT',
-      protocolVersion: 2,
-      clean: true,
-      clientId: 'my-client',
-      keepalive: 0
-    })
-    s.outStream.on('data', function (packet) {
-      t.fail('no data sent')
-    })
-    broker.on('connectionError', function (client, err) {
-      t.equal(client.version, null)
-      t.equal(err.message, 'Invalid protocol version')
-      t.equal(broker.connectedClients, 0)
-    })
-  })
+  const checkDisconnect = async () => {
+    const [client, err] = await once(s.broker, 'connectionError')
+    t.assert.equal(client.version, null)
+    t.assert.equal(err.message, 'Invalid protocol version')
+    t.assert.equal(broker.connectedClients, 0)
+  }
+
+  // run parallel
+  await Promise.all([
+    checkNoPacket(t, s),
+    checkDisconnect(),
+    sendPacket()
+  ])
 })
 
-test('reject clients that exceed the keepalive limit', function (t) {
+test('reject clients that exceed the keepalive limit', async (t) => {
   t.plan(3)
 
-  Aedes.createBroker({
-    keepaliveLimit: 100
-  }).then((broker) => {
-    t.teardown(broker.close.bind(broker))
-
-    const s = setup(broker)
-
-    s.inStream.write({
+  const s = await createAndConnect(t, {
+    broker: {
+      keepaliveLimit: 100
+    },
+    connect: {
       cmd: 'connect',
       keepalive: 150
-    })
-    s.outStream.on('data', function (packet) {
-      t.same(packet, {
-        cmd: 'connack',
-        returnCode: 6,
-        length: 2,
-        qos: 0,
-        retain: false,
-        dup: false,
-        topic: null,
-        payload: null,
-        sessionPresent: false
-      }, 'unsuccessful connack, keep alive limit exceeded')
-    })
-    broker.on('connectionError', function (client, err) {
-      t.equal(err.message, 'keep alive limit exceeded')
-      t.equal(broker.connectedClients, 0)
-    })
+    },
+    expectedReturnCode: 6
   })
+
+  const [client, err] = await once(s.broker, 'connectionError')
+  t.assert.ok(client, 'client is defined')
+  t.assert.equal(err.message, 'keep alive limit exceeded')
+  t.assert.equal(s.broker.connectedClients, 0)
 })
 
+// TODO: test fails because Aedes does not reject this
```

Comment:
is this due to changes on aedes code or race conditions? It's not clear by checking the test

Reply:
I think this was the same as the other one.